x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as kbnTestServer from 'src/core/test_helpers/kbn_server';
+
+/**
+
+1. Install a  "fake" newer version of `system`
+2. Start up Kibana
+3. Assert that the newer version of `system` is removed and the latest available version is installed instead
+
+ */
+
+describe('managed package rollbacks', () => {
```

Comment:
I spoke w/ @joshdover about using these Jest integration tests for this use case, and I think it'd be ideal. However I'm having trouble actually getting this to run. Jest seems to just not detect my test file:

```
$ yarn test:jest_integration x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts
yarn run v1.22.17
$ node scripts/jest_integration x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts
 info yarn jest x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts --runInBand --config /Users/kylepollich/Workspace/elastic/kibana/jest.config.integration.js
jest-haste-map: duplicate manual mock found: services
  The following files share their name; please delete one of them:
    * <rootDir>/src/plugins/discover/public/__mocks__/services.ts
    * <rootDir>/src/plugins/chart_expressions/expression_metric/public/__mocks__/services.ts

jest-haste-map: duplicate manual mock found: format_service
  The following files share their name; please delete one of them:
    * <rootDir>/src/plugins/chart_expressions/expression_metric/public/__mocks__/format_service.ts
    * <rootDir>/src/plugins/chart_expressions/expression_tagcloud/public/__mocks__/format_service.ts

jest-haste-map: duplicate manual mock found: index
  The following files share their name; please delete one of them:
    * <rootDir>/src/core/server/saved_objects/import/lib/__mocks__/index.ts
    * <rootDir>/packages/kbn-es-query/src/kuery/grammar/__mocks__/index.ts

No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/kylepollich/Workspace/elastic/kibana
  10245 files checked.
  roots: /Users/kylepollich/Workspace/elastic/kibana/src, /Users/kylepollich/Workspace/elastic/kibana/packages - 10245 matches
  testMatch: **/integration_tests**/*.test.{js,mjs,ts,tsx} - 89 matches
  testPathIgnorePatterns: /Users/kylepollich/Workspace/elastic/kibana/packages/kbn-ui-framework/dist/, /Users/kylepollich/Workspace/elastic/kibana/packages/kbn-pm/dist/ - 10244 matches
  testRegex:  - 0 matches
Pattern: x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts - 0 matches
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

Reply:
most of the integration tests are here, maybe you can try moving them: https://github.com/elastic/kibana/tree/main/x-pack/test/fleet_api_integration/apis

Reply:
Ah you know what, this was disabled for x-pack a while back. I can't remember why anymore. It seems to work if you add `'<rootDir>/x-pack/plugins'` to the `roots` array in `jest.config.integration.js`.

Another note is that you'll need to specify `{ oss: false }` in your `createRoot` call:

```
    root = kbnTestServer.createRoot({}, { oss: false });
```

Reply:
Thanks, Josh. I made those two changes and was able to get this integration test to run, but it looks like it's failing to boot up Kibana within the 30s timeout. 

```
 FAIL  x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts (43.889 s)
  ● managed package rollbacks › when newer version of managed package is installed › rolls back to latest supported version of package

    thrown: "Exceeded timeout of 30000 ms for a hook.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      19 |   let root: ReturnType<typeof kbnTestServer.createRoot>;
      20 |
    > 21 |   beforeAll(async () => {
         |   ^
      22 |     root = kbnTestServer.createRoot({}, { oss: false });
      23 |
      24 |     await root.preboot();

      at x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts:21:3
      at Object.<anonymous> (x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts:18:1)


  ● Test suite failed to run

    Cannot stop server while in starting phase

      223 |
      224 |       const shutdownTimeout = await this.shutdownTimeout$.pipe(take(1)).toPromise();
    > 225 |       await this.server.stop({ timeout: shutdownTimeout.asMilliseconds() });
          |                         ^
      226 |
      227 |       this.log.debug(`http server stopped`);
      228 |

      at Object.<anonymous>.module.exports.internals.Core._stop (node_modules/@hapi/hapi/lib/core.js:398:19)
      at Object.<anonymous>.internals.Server.stop (node_modules/@hapi/hapi/lib/server.js:579:27)
      at HttpServer.stop (src/core/server/http/http_server.ts:225:25)
      at HttpService.stop (src/core/server/http/http_service.ts:233:7)
      at Server.stop (src/core/server/server.ts:343:5)
      at Root.shutdown (src/core/server/root/index.ts:82:5)
      at x-pack/plugins/fleet/server/integration_tests/managed_package_rollbacks.test.ts:30:5

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        44.683 s, estimated 95 s
Ran all test suites matching /x-pack\/plugins\/fleet\/server\/integration_tests\/managed_package_rollbacks.test.ts/i.
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
error Command failed with exit code 1.
```

Would you be able to point me to any docs about getting this up and running? I did find https://github.com/elastic/kibana/blob/main/dev_docs/tutorials/testing_plugins.mdx but wondering if there's anything else about specifically setting up these tests for x-pack plugins.

Reply:
My guess is that you need to:
1. Use the `kbnTestServer.createTestServers` instead of `createRoot` and wire up it's `adjustTimeout` setting to `jest.setTimeout`
2. Start ES before starting Kibana. The return value of `createTestServers` gives you a way to do it.

Here is a good concise example of what I mean: https://github.com/elastic/kibana/blob/be61ff1d233522f6e789115107bcf874234751ea/src/core/server/saved_objects/migrations/integration_tests/migration_from_older_v1.test.ts/#L66-L110

Make sure to still supply the `oss: false` option so that X-Pack plugins are loaded.

Reply:
Thanks again Josh - this has been extremely helpful, and I've made some progress. I've been able to get Elasticsearch started up properly in the integration test suite at this point, but Kibana never actually boots, and the test eventually times out. 

I do get this logging output which makes me think I'm missing something in the `createRoot` call:

```
  i Kibana has not been configured.

    Go to http://localhost:0/?code=325235 to get started.

      at SafeSubscriber._next (src/plugins/interactive_setup/server/plugin.ts:127:19)
```